### PR TITLE
[#2497] Error messages usability improvements

### PR DESCRIPTION
**Pull request by jlucovsky**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [X] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [X] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: [2497](https://redmine.openinfosecfoundation.org/issues/2497)

Describe changes:
This PR adds a subsystem and module identifier to `SCLog` messages when the log format includes `%S`. Subsystem and module identifiers are intrinsic properties of threads and source code modules (respectively).

New threads are assigned a subsystem identifier when the thread is created; the identifier is a Thread-Local-Storage variable declared in `util-debug.c`; values are assigned to it as threads are created using `SCSetSubsystem` (a macro defined in `util-debug.h`).

Module identifiers are assigned using `SCSetModule` (macro defined in  `util-debug.h`).

Subsystem and module identifiers are added to log messages when the format contains `%S`. The generated log message will substitute a tag built from
1. The subsystem identifier
1. (Optional) The module identifier will be included if set (not all source modules will set one)
The constructed tag is of the form `[subsystem-id[:module-identifier]]` (the first and last brackets are literal; the brackets surrounding the module-identifier indicate the module-identifier is optional).

This PR was created to illustrate an approach to addressing 2497. There are some code paths where a subsystem and/or module identifier will be required and those will be addressed if this PR moves forward.

### CODE REVIEW

**Comment by victorjulien** [comment]
Just had a quick look. I like where this is going. The output does seem in a bit inaccurate in places, see:
<pre>
SC_LOG_FORMAT="%t <%d> %S: " ./src/suricata -c suricata.yaml -l tmp/ --disable-detection -r ~/sync/pcap/sandnet.pcap -v
21/12/2018 -- 07:55:21 <Info> [Suricata-main]: detection engine disabled
Error opening file /var/log/suricata/suricata.log
21/12/2018 -- 07:55:22 <Notice> [Suricata-main]: This is Suricata version 4.1.0-dev (rev 6c121a7aa)
21/12/2018 -- 07:55:22 <Info> [Suricata-main]: CPUs/cores online: 4
21/12/2018 -- 07:55:22 <Info> [Suricata-main]: Could not set core dump size to unlimited; it's set to the hard limit.
21/12/2018 -- 07:55:22 <Info> [Suricata-main]: fast output device (regular) initialized: fast.log
21/12/2018 -- 07:55:22 <Info> [Suricata-main]: eve-log output device (regular) initialized: eve.json
21/12/2018 -- 07:55:22 <Info> [Suricata-main]: stats output device (regular) initialized: stats.log
21/12/2018 -- 07:55:22 <Info> [RX#01:receive-pcap-file]: Checking file or directory /home/victor/sync/pcap/sandnet.pcap
21/12/2018 -- 07:55:22 <Info> [RX#01]: /home/victor/sync/pcap/sandnet.pcap: Plain file, not a directory
21/12/2018 -- 07:55:22 <Info> [RX#01:receive-pcap-file]: Argument /home/victor/sync/pcap/sandnet.pcap was a file
21/12/2018 -- 07:55:22 <Notice> [Detect]: all 5 packet processing threads, 4 management threads initialized, engine started.
21/12/2018 -- 07:55:22 <Info> [RX#01:receive-pcap-file]: Starting file run for /home/victor/sync/pcap/sandnet.pcap
21/12/2018 -- 07:55:22 <Info> [RX#01]: No packets with invalid checksum, assuming checksum offloading is NOT used
21/12/2018 -- 07:55:48 <Info> [RX#01:pcap-file-dispatch]: pcap file /home/victor/sync/pcap/sandnet.pcap end of file reached (pcap err code 0)
21/12/2018 -- 07:55:48 <Notice> [Detect]: Signal Received.  Stopping engine.
21/12/2018 -- 07:55:48 <Info> [Detect]: time elapsed 26.482s
21/12/2018 -- 07:55:49 <Notice> [RX#01:receive-pcap-file]: Pcap-file module read 1 files, 1354508 packets, 413760573 bytes
</pre>
E.g. "21/12/2018 -- 07:55:48 <Notice> [Detect]: Signal Received.  Stopping engine." here we should see 'Detect' but probably 'Suricata-Main'.

Also, in "21/12/2018 -- 07:55:22 <Info> [RX#01]: No packets with invalid checksum, assuming checksum offloading is NOT used" here it should probably be 'RX#01:receive-pcap-file" or something similar.

There are some issues with the patch. For it to compile I had to do these fixes:
```diff
diff --git a/src/util-debug.c b/src/util-debug.c
index 216165e33..b084a2d07 100644
--- a/src/util-debug.c
+++ b/src/util-debug.c
@@ -210,7 +210,7 @@ static inline void SCLogPrintToSyslog(int syslog_log_level, const char *msg)
  */
 static int SCLogMessageJSON(struct timeval *tval, char *buffer, size_t buffer_size,
         SCLogLevel log_level, const char *file,
-        unsigned line, const char *function, SCError error_code,
+        unsigned line, const char *function, const char *module, SCError error_code,
         const char *message)
 {
     json_t *js = json_object();
diff --git a/src/util-debug.h b/src/util-debug.h
index 8ec5079c8..efa322fbe 100644
--- a/src/util-debug.h
+++ b/src/util-debug.h
@@ -214,7 +214,7 @@ static const char *_sc_module;
 /* The subsystem name, usually the thread's name */
 #define SCSetSubsystem(subsystem_name)                                          \
     do {                                                                        \
-        __thread extern const char *_sc_subsystem;                              \
+        extern __thread const char *_sc_subsystem;                              \
         _sc_subsystem = subsystem_name;                                         \
     } while(0)
```
Note that SCLogMessageJSON would need to actually support the module as well.

More serious is this runtime ASAN error:
```
21/12/2018 -- 07:55:49 <Notice> [RX#01:receive-pcap-file]: Pcap-file module read 1 files, 1354508 packets, 413760573 bytes
=================================================================
==27485==ERROR: AddressSanitizer: heap-use-after-free on address 0x6020002fd350 at pc 0x7fe7eeeb0075 bp 0x7fff9e368e70 sp 0x7fff9e3685e8
READ of size 7 at 0x6020002fd350 thread T0 (Suricata-Main)
    #0 0x7fe7eeeb0074  (/usr/lib/x86_64-linux-gnu/libasan.so.4+0x74074)
    #1 0x7fe7eef2555f  (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xe955f)
    #2 0x7fe7eeeb01ed  (/usr/lib/x86_64-linux-gnu/libasan.so.4+0x741ed)
    #3 0x7fe7eeedcf4d in __interceptor_vsnprintf (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xa0f4d)
    #4 0x7fe7eeedd286 in snprintf (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xa1286)
    #5 0x56049586fed0 in SCLogMessageGetBuffer /home/victor/devel/eidps/src/util-debug.c:461
    #6 0x560495870a9a in SCLogMessage /home/victor/devel/eidps/src/util-debug.c:601
    #7 0x5604949a76c2 in StatsLogSummary /home/victor/devel/eidps/src/counters.c:832
    #8 0x5604949ac44d in StatsReleaseResources /home/victor/devel/eidps/src/counters.c:1242
    #9 0x56049580129d in PostRunDeinit /home/victor/devel/eidps/src/suricata.c:2329
    #10 0x560495808d13 in main /home/victor/devel/eidps/src/suricata.c:3035
    #11 0x7fe7ebb76b96 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b96)
    #12 0x56049456e8a9 in _start (/home/victor/sync/devel/eidps/src/suricata+0x2518a9)

0x6020002fd350 is located 0 bytes inside of 7-byte region [0x6020002fd350,0x6020002fd357)
freed by thread T0 (Suricata-Main) here:
    #0 0x7fe7eef1a7b8 in __interceptor_free (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xde7b8)
    #1 0x5604958275a5 in TmThreadFree /home/victor/devel/eidps/src/tm-threads.c:1827
    #2 0x560495828085 in TmThreadClearThreadsFamily /home/victor/devel/eidps/src/tm-threads.c:1878
    #3 0x560495801293 in PostRunDeinit /home/victor/devel/eidps/src/suricata.c:2323
    #4 0x560495808d13 in main /home/victor/devel/eidps/src/suricata.c:3035
    #5 0x7fe7ebb76b96 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b96)

previously allocated by thread T0 (Suricata-Main) here:
    #0 0x7fe7eeeb3538 in strdup (/usr/lib/x86_64-linux-gnu/libasan.so.4+0x77538)
    #1 0x560495827829 in TmThreadSetGroupName /home/victor/devel/eidps/src/tm-threads.c:1855
    #2 0x560495574595 in RunModeFilePcapAutoFp /home/victor/devel/eidps/src/runmode-pcap-file.c:252
    #3 0x5604955860a1 in RunModeDispatch /home/victor/devel/eidps/src/runmodes.c:384
    #4 0x560495808a86 in main /home/victor/devel/eidps/src/suricata.c:3002
    #5 0x7fe7ebb76b96 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b96)

SUMMARY: AddressSanitizer: heap-use-after-free (/usr/lib/x86_64-linux-gnu/libasan.so.4+0x74074) 
Shadow bytes around the buggy address:
  0x0c0480057a10: fa fa fd fd fa fa 01 fa fa fa 01 fa fa fa 00 06
  0x0c0480057a20: fa fa 00 06 fa fa 00 00 fa fa 00 00 fa fa 00 00
  0x0c0480057a30: fa fa 00 00 fa fa 00 07 fa fa 00 00 fa fa 07 fa
  0x0c0480057a40: fa fa fd fd fa fa 00 fa fa fa fd fa fa fa 00 fa
  0x0c0480057a50: fa fa fd fd fa fa 00 fa fa fa 00 fa fa fa fd fa
=>0x0c0480057a60: fa fa fd fa fa fa fd fa fa fa[fd]fa fa fa fa fa
  0x0c0480057a70: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0480057a80: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0480057a90: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0480057aa0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0480057ab0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==27485==ABORTING
```

Finally, please have a look at the https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Code_Submission_Quality_Criteria
The commit message especially needs a bit of love.

Thanks!

File: src/alert-prelude.c [hunk 1]
```diff
@@ -1036,6 +1036,8 @@ static TmEcode AlertPreludeThreadInit(ThreadVars *t, const void *initdata, void
         SCReturnInt(TM_ECODE_FAILED);
     }
 
+    printf(">>>>>>>>>> [%s] %s <<<<<<<<<<<\n", __func__, t->name);
```
Comment (victorjulien) [inline src/alert-prelude.c]: minor: left over from debugging

File: src/util-debug.c [hunk 2]
```diff
@@ -445,6 +448,29 @@ static SCError SCLogMessageGetBuffer(
                 substr++;
                 break;
 
+            case SC_LOG_FMT_SUBSYSTEM:
+                temp_fmt[0] = '\0';
+                const char *fmt_string = "%s%s%s%s";
+                if (_sc_subsystem || module) {
+                    if (_sc_subsystem && module) {
+                        fmt_string = "%s%s[%s:%s]%s";
```
Comment (victorjulien) [inline src/util-debug.c]: I think the [ ] should not be part of this here. It can be added in the default and/or user supplied format string instead.

**Comment by victorjulien** [comment]
The use-after-free ASAN reports is probably due to the thread data structures being freed, which includes the thread names, and then some logging happening after that. So that'd leave _sc_subsystem as a dangling ptr?

File: src/alert-prelude.c (hunk 1 above)
Comment (jlucovsky) [reply src/alert-prelude.c]: Removed.

File: src/util-debug.c (hunk 2 above)
Comment (jlucovsky) [reply src/util-debug.c]: Will move to the default log formats.

Comment (jlucovsky) [reply src/util-debug.c]: Brackets moved to default log format string.

**Comment by jlucovsky** [comment]
Running with ASAN no longer produces a crash when the program terminates. Suricata was built with ` -fsanitize=address -fno-omit-frame-pointer` added to CFLAGS for this test.

**Comment by victorjulien** [comment]
Please implement changes in a new PR. See https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Github_work_flow#Create-a-new-branch-for-incorporating-feedback

**Pull request closed without merging**